### PR TITLE
ggml : remove bit shuffling

### LINKED ISSUE #1241: Use different bit arrangement for quants (nibbles)
In the existing `llama.cpp` implementation, quantization bits of consecutive model weights are packed together one after the other. E.g., for 4-bit quantization, the 8 bits of two consecutive weights are stored into a `uint8_t`. The disadvantage of this approach is that when the data is to be used in dot products or is being de-quantized for matrix multiplications done via BLAS, and the operations are performed using SIMD instructions, one needs to shuffle the de-quantized bytes to get them into the correct order. These shuffle operations can be avoided by arranging the bits differently. For instance, for 4-bit quantization in blocks of 32 weights (`Q4_0`), one can store the quants of the first 16 weights into the low 4 bits of the 16 `uint8_t`'s, and the quants of the second 16 weights in the block of 32 into the high 4-bits. The same or similar strategy can also be applied for other block sizes or when using 2 bits per weight. 

The performance gain is not earth-shattering: in a synthetic benchmark performing `Q4_0_Q8_0` dot products I measured about a 10% speedup from avoiding the shuffle. Still, it is a trivial change, so why leave this low-hanging fruit hanging?

### PR DESCRIPTION
**Pull request by ggerganov**
Close #1241 

- Drop `Q4_2` support
- Changed bit-order for `Q4` and `Q5` (breaking change)
- Preplexity is perplexing as usual

---

New timings:

| Model | Measure      | F16    | Q4_0   | Q4_1   | Q5_0   | Q5_1   | Q8_0   |
|------:|--------------|-------:|-------:|-------:|-------:|-------:|-------:|
|    7B | ms/tok @ 4th |    128 |     50 |     54 |     75 |     83 |     75 |
|    7B | ms/tok @ 8th |    123 |     44 |     52 |     53 |     58 |     72 |
|   13B | ms/tok @ 4th |    332* |    93 |    101 |    150 |    164 |    141 |
|   13B | ms/tok @ 8th |    308* |     81 |     96 |    96 |    104 |    136 |

* *these numbers vary a lot since the model is on the 32GB limit of my MacBook*

Old timings:

| Model | Measure      | F16    | Q4_0   | Q4_1   | Q5_0   | Q5_1   | Q8_0   |
|------:|--------------|-------:|-------:|-------:|-------:|-------:|-------:|
|    7B | ms/tok @ 4th |    128 |     56 |     61 |     91 |     95 |     75 |
|    7B | ms/tok @ 8th |    128 |     47 |     55 |     53 |     59 |     75 |
|   13B | ms/tok @ 4th |    239 |    104 |    113 |    176 |    185 |    141 |
|   13B | ms/tok @ 8th |    240 |     85 |     99 |    108 |    117 |    147 |

*overall, all these numbers seem to have about +/- 10% variablility from run to run. not ideal benchmark, but not sure what else to do*

### CODE REVIEW

**Comment by ggerganov** [comment]
Ah, that's unfortunate. I don't think there's a way to fix this efficiently for the `ARM_NEON` branch

File: ggml.c [hunk 1]
```diff
@@ -828,664 +734,188 @@ static_assert(sizeof(block_q8_0) == sizeof(float) + QK8_0, "wrong q8_0 block siz
...
-            const v128_t vi = wasm_i32x4_trunc_sat_f32x4(vf);
-            const v128_t vc = wasm_i32x4_min(vi, wasm_i32x4_splat(15));
-
-            y[i].qs[2*l + 0] = wasm_i32x4_extract_lane(vc, 0) | (wasm_i32x4_extract_lane(vc, 1) << 4);
-            y[i].qs[2*l + 1] = wasm_i32x4_extract_lane(vc, 2) | (wasm_i32x4_extract_lane(vc, 3) << 4);
-        }
-    }
-#else
-    // scalar
+static void quantize_row_q4_0(const float * restrict x, void * restrict y, int k) {
```
[the remark this answers is not in the document]
Comment (ggerganov) [reply ggml.c:781]: Yes, will refactor this after we merge this stuff

File: README.md [hunk 2]
```diff
@@ -7,18 +7,10 @@
...
-
-**The quantization formats will soon be updated: https://github.com/ggerganov/llama.cpp/pull/1305**
-
-**All `ggml` model files using the old format will not work with the latest `llama.cpp` code after that change is merged**
-
----
-
 **Hot topics:**
 
+- Qauntization formats `Q4` and `Q5` have changed - requantize any old models [(info)](https://github.com/ggerganov/llama.cpp/pull/1405)
```
Comment (Excigma) [inline README.md:12]: Is "qauntization" a typo? :thinking:

**Comment by redthing1** [comment]
Is there a script to upgrade the old models to new? I don't have the source models because they're huge.

**Comment by ggerganov** [comment]
#1384 does not work for NEON because when we remove the `vzip` calls, the high bits are no longer in the right place.

This is the relevant section before this PR:

https://github.com/ggerganov/llama.cpp/blob/b608b55a3ea8e4760c617418538465449175bdb8/ggml.c#L3335-L3360

We were ORing the 5th bit after the `vzip`.
I didn't see a way to fix this without shuffling the tables or the bits in some way.

**Comment by M00N-MAN** [comment]
> Is there a script to upgrade the old models to new? I don't have the source models because they're huge.

Hello All

Could someone please share the way of requantizing? 
image
At least on this example https://huggingface.co/TheBloke/WizardLM-7B-uncensored-GGML/blob/main/README.md
What should be done with WizardLM-7B-uncensored.q5_1.bin or which tool need to by used for https://huggingface.co/ehartford/WizardLM-13B-Uncensored which is the source model.

**Comment by LostRuins** [comment]
Cool to see this is merged, i'm slightly confused though, seems like the converter is still writing the old file version (1) ? 
image

**Comment by philpax** [comment]
> Cool to see this is merged, i'm slightly confused though, seems like the converter is still writing the old file version (1) ? 
> image
> 

I think that's intentional - the converter converts the f16 model to a GGJT v1 f16 model, which is then quantised to a GGJT v2 qX_Y model. (The saver always writes version 2, from what I can see.)

**Comment by LostRuins** [comment]
Doesn't seem like it though, I don't see references where File version 2 is set during quantization either.

**Comment by TheBloke** [comment]
> Could someone please share the way of requantizing? <img alt="image" width="594" src="https://user-example https://huggingface.co/TheBloke/WizardLM-7B-uncensored-GGML/blob/main/README.md What should be done with WizardLM-7B-uncensored.q5_1.bin or which tool need to by used for https://huggingface.co/ehartford/WizardLM-13B-Uncensored which is the source model.

Check my repos again. I've re-quantised all my GGMLs using the latest code, in q4_0, q5_0, q5_1 and q8_0 variants.

**Comment by KerfuffleV2** [comment]
Is losslessly converting from the old format to the new one possible? i.e. someone just needs to write a tool, or are there changes here that make conversion impossible?

If it is possible, can anyone explain exactly how the data format changed? I can probably write a conversion tool with that information.

**Review by BobbyKay1 — APPROVED** [review]

**Pull request merged**